ironfish/src/blockchain/blockchain.test.ts

```diff
@@ -164,6 +164,55 @@ describe('Blockchain', () => {
     expect(blocks[0].hash.equals(genesis.hash)).toBe(true)
   })
 
+  it('should not iterate and jump chains', async () => {
+    const { strategy, chain } = nodeTest
+    strategy.disableMiningReward()
+
+    const genesis = chain.genesis
+
+    // This test checks that when iterating a reorg is happening, we don't
+    // suddenly jump chains while the table is being re-written when we don't
+    // have a lock on the main chain table. We try to iterate to A2, then do a
+    // reorg and see if the next iteration incorrectly yields B3.
+    //
+    // G -> A1 -> A2 -> A3
+    //         -> B2 -> B3 -> B4
+
+    const blockA1 = await makeBlockAfter(chain, genesis)
+    const blockA2 = await makeBlockAfter(chain, blockA1)
+    const blockA3 = await makeBlockAfter(chain, blockA2)
+    const blockB2 = await makeBlockAfter(chain, blockA1)
+    const blockB3 = await makeBlockAfter(chain, blockB2)
+    const blockB4 = await makeBlockAfter(chain, blockB3)
+
+    await expect(chain).toAddBlock(blockA1)
+    await expect(chain).toAddBlock(blockA2)
+    await expect(chain).toAddBlock(blockA3)
+
+    expect(chain.head.hash.equals(blockA3.header.hash)).toBe(true)
+    expect(chain.latest.hash.equals(blockA3.header.hash)).toBe(true)
+
+    const iter = chain.iterateTo(genesis, blockA2.header)
```

Comment:
I think this should iterate to A3, or else I'd expect the iterator will always be done when yielding A2 regardless of whether a fork was added.

Reply:
Fixed by making tests more robust